### PR TITLE
:sparkles: Rewrite notifications and implement APIv1 endpoint

### PR DESCRIPTION
**Pull request by MrKrisKrisu**
In this PR I want to rewrite our usage of laravels notification components, remove the frontend endpoint, the backend rendered html and make our notifications api at v1 working correctly.

### CODE REVIEW

**Comment by MrKrisKrisu** [comment]
Hi (at)App-Devs! This is the "new structured" Notification-Endpoint in APIv1. Is there anything you need extra or are you happy with the structure? There is still space for correction. :)

(cc @jheubuch @vainamov)

# Request

````
Request URL: http://localhost:8000/api/v1/notifications
Request Method: GET
Accept: application/json
````

# (Empty) Response

````json
{
    "data": [
        
    ],
    "links": {
        "first": "http://localhost:8000/api/v1/notifications?page=1",
        "last": null,
        "prev": null,
        "next": null
    },
    "meta": {
        "current_page": 1,
        "from": 1,
        "path": "http://localhost:8000/api/v1/notifications",
        "per_page": 15,
        "to": 7
    }
}
````


# Accepted event
```json
{
    "id": "5034f0a1-887c-4765-b651-70e611440622",
    "type": "EventSuggestionProcessed",
    "leadFormatted": "Your event suggestion <b>Gulaschprogrammiernacht</b> was reviewed.",
    "lead": "Your event suggestion Gulaschprogrammiernacht was reviewed.",
    "noticeFormatted": "Your suggestion was accepted.",
    "notice": "Your suggestion was accepted.",
    "link": "http://localhost:8000/statuses/event/gulaschprogrammiernacht1",
    "data": {
        "accepted": true,
        "event": {
            "id": 9,
            "slug": "gulaschprogrammiernacht1",
            "name": "Gulaschprogrammiernacht",
            "begin": "2023-06-07T22:00:00.000000Z",
            "end": "2023-06-11T21:59:59.000000Z"
        },
        "suggestedName": "Gulaschprogrammiernacht"
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:10:50+02:00",
    "createdAtForHumans": "2 seconds ago"
}
```

# Denied Event

```json
{
    "id": "ff843ed0-4559-4f54-a682-25be1cc36134",
    "type": "EventSuggestionProcessed",
    "leadFormatted": "Your event suggestion <b>Stupid Event</b> was reviewed.",
    "lead": "Your event suggestion Stupid Event was reviewed.",
    "noticeFormatted": "Your suggestion was rejected.",
    "notice": "Your suggestion was rejected.",
    "link": null,
    "data": {
        "accepted": false,
        "event": null,
        "suggestedName": "Stupid Event"
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:06:35+02:00",
    "createdAtForHumans": "3 minutes ago"
}
```

# Follow request
```json
{
    "id": "d611df07-510d-4eca-b8db-91934ca65abe",
    "type": "FollowRequestIssued",
    "leadFormatted": "<b>@bob</b> wants to follow you.",
    "lead": "@bob wants to follow you.",
    "noticeFormatted": "To accept or decline the follow request, click here or go to your settings.",
    "notice": "To accept or decline the follow request, click here or go to your settings.",
    "link": "http://localhost:8000/settings/follower",
    "data": {
        "followRequest": {
            "id": 5
        },
        "user": {
            "id": 2,
            "username": "bob",
            "name": "Bob from Alice & Bob"
        }
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:09:08+02:00",
    "createdAtForHumans": "48 seconds ago"
}
```

# Approved Follow request
```json
{
    "id": "21280117-cdf3-47bc-bc6f-e727c961ffc5",
    "type": "FollowRequestApproved",
    "leadFormatted": "<b>@Gertrud123</b> has approved your follow request",
    "lead": "@Gertrud123 has approved your follow request",
    "noticeFormatted": null,
    "notice": "",
    "link": "http://localhost:8000/@Gertrud123",
    "data": {
        "follow": {
            "id": 9
        },
        "user": {
            "id": 1,
            "username": "Gertrud123",
            "name": "Gertrud"
        }
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:11:29+02:00",
    "createdAtForHumans": "3 seconds ago"
}
```

# Status liked
```json
{
    "id": "10670573-c08f-4bb0-a2d9-12ceba6aafba",
    "type": "StatusLiked",
    "leadFormatted": "<b>@bob</b> liked your check-in.",
    "lead": "@bob liked your check-in.",
    "noticeFormatted": "Journey in S 8 on 2023-06-06",
    "notice": "Journey in S 8 on 2023-06-06",
    "link": "http://localhost:8000/status/23",
    "data": {
        "like": {
            "id": 4
        },
        "status": {
            "id": 23
        },
        "trip": {
            "origin": {
                "id": 251,
                "ibnr": 8079041,
                "name": "Karlsruhe Bahnhofsvorplatz"
            },
            "destination": {
                "id": 292,
                "ibnr": 8000110,
                "name": "Freudenstadt Hbf"
            },
            "plannedDeparture": "2023-06-06T12:16:56.000000Z",
            "plannedArrival": "2023-06-06T14:07:00.000000Z",
            "lineName": "S 8"
        },
        "liker": {
            "id": 2,
            "username": "bob",
            "name": "Bob from Alice & Bob"
        }
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:17:23+02:00",
    "createdAtForHumans": "53 seconds ago"
}
```

# User followed
```json
{
    "id": "ebdd8f66-9ab4-448d-ad40-e8486f0570a1",
    "type": "UserFollowed",
    "leadFormatted": "<b>@bob</b> is following you.",
    "lead": "@bob is following you.",
    "noticeFormatted": null,
    "notice": "",
    "link": "http://localhost:8000/@bob",
    "data": {
        "follow": {
            "id": 10
        },
        "follower": {
            "id": 2,
            "username": "bob",
            "name": "Bob from Alice & Bob"
        }
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:19:19+02:00",
    "createdAtForHumans": "2 seconds ago"
}
```

# User joined connection
```json
{
    "id": "bb1ba9a5-9c2b-43c3-b8c9-2f70651fc51c",
    "type": "UserJoinedConnection",
    "leadFormatted": "<b>@bob</b> is in your connection!",
    "lead": "@bob is in your connection!",
    "noticeFormatted": "@bob is on <b>S 8</b> from <b>Karlsruhe Bahnhofsvorplatz</b> to <b>Freudenstadt Hbf</b>.",
    "notice": "@bob is on S 8 from Karlsruhe Bahnhofsvorplatz to Freudenstadt Hbf.",
    "link": "http://localhost:8000/status/24",
    "data": {
        "status": {
            "id": 24
        },
        "checkin": {
            "linename": "S 8",
            "origin": "Karlsruhe Bahnhofsvorplatz",
            "destination": "Freudenstadt Hbf"
        },
        "user": {
            "id": 2,
            "username": "bob",
            "name": "Bob from Alice & Bob"
        }
    },
    "readAt": null,
    "createdAt": "2023-06-06T14:19:48+02:00",
    "createdAtForHumans": "2 seconds ago"
}
```

**Comment by vainamov** [comment]
Considering the vast majority (if not all?) endpoints return property names with lower camel casing I would prefer that to be consistent. So `leadFormatted`/`noticeFormatted` (and `currentPage`/`perPage` for that matter, but this is already part of the API so...) seem more appropriate to me.

The structure of the reponses looks intuitive. The only thing that stood out to me is the presence of `icon` and `link`, which (at least for our application) doesn't seem helpful, but I assume this is used for the Träwelling UI and wouldn't bother me if it's kept in the response.

**Comment by MrKrisKrisu** [comment]
> Considering the vast majority (if not all?) endpoints return property names with lower camel casing I would prefer that to be consistent. So leadFormatted/noticeFormatted (and currentPage/perPage for that matter, but this is already part of the API so...) seem more appropriate to me.

Well. Thank you for noticing that. I will change that! :)

> The structure of the reponses looks intuitive. The only thing that stood out to me is the presence of icon and link, which (at least for our application) doesn't seem helpful, but I assume this is used for the Träwelling UI and wouldn't bother me if it's kept in the response.

Will try to remove `icon`. :)
In my opinion the link can be a good marker for apps, too, if they catch the "traewelling.de" locally.

**Comment by jheubuch** [comment]
First of all, thanks for your effort! ❤️

I just got a small remark, could we have more details about the trip in the `StatusLiked` notification? 

> "trip": {
            "departure": "2023-06-06T12:16:56.000000Z",
            "linename": "S 8"
        }

Maybe we can get an `origin` and a `destination` just like in `UserJoinedConnection`?

**Comment by MrKrisKrisu** [comment]
> Maybe we can get an `origin` and a `destination` just like in `UserJoinedConnection`?

Sure! I don't want to expose the whole status/trip here, but these attributes make sense for me. :)

**Comment by MrKrisKrisu** [comment]
> Sure! I don't want to expose the whole status/trip here, but these attributes make sense for me. :)

@jheubuch Please have a look at the updated examples above - okay for you?

**Comment by jheubuch** [comment]
> @jheubuch Please have a look at the updated examples above - okay for you?

Wonderful!

**Comment by vainamov** [comment]
Looks awesome, thank you for the improvements ❤️

File: .github/workflows/phpunit.yml [hunk 1]
```diff
@@ -47,4 +47,4 @@ jobs:
         env:
           DB_CONNECTION: sqlite
           DB_DATABASE: database/database.sqlite
-        run: php artisan test --parallel --coverage --min=44 #Increase this value regularly while writing tests. Target: 80%
+        run: php artisan test --parallel --coverage --min=43 #Increase this value regularly while writing tests. Target: 80%
```
Comment (HerrLevin) [inline .github/workflows/phpunit.yml:50]: 👀

File: app/Http/Controllers/API/v1/Controller.php [hunk 2]
```diff
@@ -84,9 +86,9 @@
 class Controller extends \App\Http\Controllers\Controller
 {
     public function sendResponse(
-        array|string|object $data = null,
-        int                 $code = 200,
-        array               $additional = null
+        array|string|int|object $data = null,
```
Comment (HerrLevin) [inline app/Http/Controllers/API/v1/Controller.php]: maybe even `mixed`?

File: app/Http/Controllers/API/v1/NotificationsController.php [hunk 3]
```diff
@@ -11,60 +11,167 @@
...
 
     /**
-     * Get all latest Messages
+     * @OA\Get(
+     *      path="/notifications",
+     *      operationId="listNotifications",
+     *      tags={"Notifications"},
+     *      summary="Get paginated notifications for authenticated user",
+     *      description="Returns paginated notifications of a authenticated",
+     *      @OA\Response(
```
Comment (HerrLevin) [inline app/Http/Controllers/API/v1/NotificationsController.php:50]: Response 401

File: app/Http/Controllers/API/v1/NotificationsController.php [hunk 4]
```diff
@@ -11,60 +11,167 @@
...
     /**
-     * Get the amount of (unread) messages
-     * @return JsonResponse
+     * @OA\Get(
+     *      path="/notifications/unread/count",
+     *      operationId="getUnreadCount",
+     *      tags={"Notifications"},
+     *      summary="Get count of unread notifications for authenticated user",
+     *      description="Returns count of unread notifications of a authenticated user",
+     *      @OA\Response(
```
Comment (HerrLevin) [inline app/Http/Controllers/API/v1/NotificationsController.php:19]: Response 401

File: app/Http/Controllers/API/v1/NotificationsController.php [hunk 5]
```diff
@@ -11,60 +11,167 @@
...
+     *      tags={"Notifications"},
+     *      summary="Mark notification as read",
+     *      @OA\Parameter (
+     *          name="id",
+     *          in="path",
+     *          description="ID of notification",
+     *          example="cbf6054e-9c00-4b1f-ab37-7eb18ac8419f",
+     *          @OA\Schema(type="string")
+     *      ),
+     *      @OA\Response(
```
Comment (HerrLevin) [inline app/Http/Controllers/API/v1/NotificationsController.php:91]: Response 401

File: resources/views/includes/modals/notifications-board.blade.php [hunk 6]
```diff
@@ -6,16 +6,18 @@
...
                 </a>
                 <button type="button" class="btn-close" data-mdb-dismiss="modal" aria-label="Close"></button>
             </div>
             <div class="modal-body" id="notifications-list">
                 <div id="notifications-empty" class="text-center text-muted">
-                    {{ __('notifications.empty') }}
-                    <br/>¯\_(ツ)_/¯
+                    <div class="spinner-border text-primary" role="status">
+                        <span class="visually-hidden">Loading...</span>
+                    </div>
```
Comment (HerrLevin) [inline resources/views/includes/modals/notifications-board.blade.php:20]: Add Text for "no notifications found"

File: routes/web.php [hunk 7]
```diff
@@ -224,15 +223,6 @@
     Route::get('/trains/setHome/', [FrontendTransportController::class, 'setTrainHome'])
          ->name('user.setHome');
 
-    Route::get('/notifications/latest', [NotificationController::class, 'renderLatest'])
```
Comment (HerrLevin) [inline routes/web.php:227]: one less api-like frontend endpoint! I love it! 🥰

File: storage/api-docs/api-docs.json [hunk 8]
```diff
@@ -18,7 +18,7 @@
             "description": "Production Server"
         },
         {
-            "url": "https://traewelling.de/api/v1",
+            "url": "http://localhost:8000/api/v1",
```
Comment (HerrLevin) [inline storage/api-docs/api-docs.json:21]: 🤨🕶️🤏

**Review by HerrLevin — CHANGES_REQUESTED** [review]

**Review by HerrLevin — APPROVED** [review]

**Pull request merged**